### PR TITLE
Encode enough to make URI happy.

### PR DESCRIPTION
**Pull request by swankjesse**
Plus a bunch of test cases around this unfortunate case.

Closes https://github.com/square/okhttp/issues/1872

### CODE REVIEW

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 1]
```diff
@@ -324,19 +327,17 @@ public URL url() {
...
+   * Returns this URL as a {@link URI java.net.URI}. Because {@code URI} forbids certain characters
+   * like {@code [} and {@code |}, the returned URI may escape more characters than this URL.
+   *
+   * <p>This method throws an unchecked {@link IllegalStateException} if it cannot be converted to a
+   * URI even after escaping forbidden characters. In particular, URLs that contain malformed
+   * percent escapes like {@code http://host/%xx} will trigger this exception.
    */
   public URI uri() {
     try {
-      String uriUserInfo = username + ":" + password;
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: This was pretty broken, particularly around usernames and passwords containining characters like `:` and `@`.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 2]
```diff
@@ -867,6 +864,31 @@ public Builder encodedFragment(String encodedFragment) {
...
+     * Re-encodes the components of this URL so that it satisfies (obsolete) RFC 2396, which is
+     * particularly strict for certain components.
+     */
+    Builder reencodeForUri() {
+      for (int i = 0, size = encodedPathSegments.size(); i < size; i++) {
+        String pathSegment = encodedPathSegments.get(i);
+        encodedPathSegments.set(i,
+            canonicalize(pathSegment, PATH_SEGMENT_ENCODE_SET_URI, true, false));
+      }
+      if (encodedQueryNamesAndValues != null) {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: yep, this is a possibly-null list that has possibly-null values. that's the best I can do!

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 3]
```diff
@@ -196,11 +196,20 @@
...
-        HttpUrl.parse("http://foo:pass1@bar:pass2@baz/path"));
+    HttpUrl httpUrl = HttpUrl.parse("http://foo@bar@baz/path");
+    assertEquals("foo@bar", httpUrl.username());
+    assertEquals("", httpUrl.password());
+    assertEquals(HttpUrl.parse("http://foo%40bar@baz/path"), httpUrl);
+  }
+
+  /** Given multiple ':' characters, the first one is the delimiter. */
+  @Test public void authorityWithMultipleColons() throws Exception {
+    HttpUrl httpUrl = HttpUrl.parse("http://foo:pass1@bar:pass2@baz/path");
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Yikes... this is some straight-up phishing URL trickery

**Comment by JakeWharton** [comment]
LGTM

**Comment by JakeWharton** [comment]
```
Tests run: 226, Failures: 0, Errors: 1, Skipped: 8, Time elapsed: 15.475 sec <<< FAILURE! - in com.squareup.okhttp.URLConnectionTest
invalidHost(com.squareup.okhttp.URLConnectionTest)  Time elapsed: 0.012 sec  <<< ERROR!
java.net.UnknownHostException: 1234.1.1.1: Name or service not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:901)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1295)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1248)
	at java.net.InetAddress.getAllByName(InetAddress.java:1164)
	at java.net.InetAddress.getAllByName(InetAddress.java:1098)
	at com.squareup.okhttp.internal.Network$1.resolveInetAddresses(Network.java:29)
	at com.squareup.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:187)
	at com.squareup.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:156)
	at com.squareup.okhttp.internal.http.RouteSelector.next(RouteSelector.java:98)
	at com.squareup.okhttp.internal.http.HttpEngine.createNextConnection(HttpEngine.java:345)
	at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:328)
	at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:246)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:446)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:118)
	at com.squareup.okhttp.URLConnectionTest.invalidHost(URLConnectionTest.java:396)
```

**Comment by swankjesse** [comment]
The `UnknownHostException` exception is surprising. Looks like a bug in the test. Fixed.

**Pull request merged**